cmake: Extracted Zephyr module processing into python script

tejlmand opened the pull request:
Fixes: #14513
Fixes: #14989

This commit move the functionality of extracting zephyr modules into
generated CMake and Kconfig include files from CMake into python.

This allows other tools, especially CI to re-use the zephyr module
functionality.

tejlmand committed:
cmake: Extracted Zephyr module processing into python script

Fixes: #14513

This commit move the functionality of extracting zephyr modules into
generated CMake and Kconfig include files from CMake into python.

This allows other tools, especially CI to re-use the zephyr module
functionality.

Signed-off-by: Torsten Rasmussen <torsten.rasmussen@nordicsemi.no>

carlescufi commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,148 @@
+        with open(module_yml, 'r') as f:
+            meta = yaml.safe_load(f.read())
+
+        schema = yaml.safe_load(METADATA_SCHEMA)
```
You might need a `Loader=` param to avoid warnings with newer versions of PyYaml: https://github.com/zephyrproject-rtos/ci-tools/commit/a5298663078163a818fcf7d2c04e379050b56057

tejlmand replied:
Thanks, will add.

tejlmand replied:
Took a look at: zephyrproject-rtos/ci-tools@a529866
That code uses `yaml.load` and not `yaml.safe_load` as this code.

So that is not an issue here (as well as safe_load doesn't support the `Loader` argument)

mbolivar replied:
It also feels a bit strange to parse this schema for each module, since it isn't changing. It could either be a global variable or passed as an argument instead to avoid that.

tejlmand replied:
@mbolivar fixed

carlescufi reviewed (changes requested):
I am uneasy about this. While parsing the modules with Python seems like a good idea, I am not sure about generating CMake files from Python. This is a bit extreme, isn't it? Wouldn't it be better to print a list of paths to `CMakeLists.txt` from `zephyr_modules.py` and then have CMake parse that output and call `add_subdirectory()` on each?

tejlmand commented:
Thanks for the comments.
Let me give the reasoning for this.

> I am uneasy about this. While parsing the modules with Python seems like a good idea, I am not sure about generating CMake files from Python. This is a bit extreme, isn't it?

It was done this way, as to have same principle used for both Kconfig and CMake.
That is, generating a file that can be included by the corresponding tool.

Secondly, as commented by @SebastianBoe output from scripts should ideally go through a file, and thus, instead of generating text file with a list of folders to include, which then must be parsed by CMake, then I decided to generate the CMake file directly.
See https://github.com/zephyrproject-rtos/zephyr/pull/13067#discussion_r254696304

Third, if we in future needs to create additional type of files that can be included, e.g. DTS, then that can be supported with `--dts-out`.

The `--cmake-out` can be omitted by CI, and just use `--kconfig-out`, which means CI doesn't need to generate the CMake files for Kconfig verification.

> Wouldn't it be better to print a list of paths to `CMakeLists.txt` from `zephyr_modules.py` and then have CMake parse that output and call `add_subdirectory()` on each?

as mentioned, printing directly from scripts is discouraged, so would need to go through a file anyway.

SebastianBoe commented:
> as mentioned, printing directly from scripts is discouraged, so would need to go through a file anyway.

You don't have to print from a script to do this. You can have the python script write a file that is easily parse-able by CMake, e.g csv, or lines of set(FOO bar).

Generally, it is ok to generate data, but not code. There is a fine line between them, but e.g. add_subdirectory looks like code to me.

tejlmand commented:
> > as mentioned, printing directly from scripts is discouraged, so would need to go through a file anyway.
> 
> You don't have to print from a script to do this. You can have the python script write a file that is easily parse-able by CMake, e.g csv, or lines of set(FOO bar).
> 

Yes, that's exactly what i'm stating.
The printing was suggested by @carlescufi:
> > > Wouldn't it be better to print a list

And to that I replied a file should be generated from the Python script in any case, just as you also state.

> Generally, it is ok to generate data, but not code. There is a fine line between them, but e.g. add_subdirectory looks like code to me.

As does the Kconfig generated file:
`osource "<path>/Kconfig"`

I think the `add_subdirectory()` and `osource` are similar in nature.
But if it preferred that the CMake output generates a plain file and Kconfig output generates code with `osource` statements, then I can change.

Remember, the python script is now generic, so that it can be used outside CMake, e.g. for CI.

SebastianBoe commented:
```
As does the Kconfig generated file:
osource "<path>/Kconfig"
```

Yeah, but that's unavoidable AFAICT. But for CMake, we are able to avoid more code generation than in the current proposal AFAICT.

SebastianBoe commented:
> But if it preferred that the CMake output generates a plain file and Kconfig output generates code with osource statements, then I can change.

Yes, this is preferable.

carlescufi commented:
I see that @SebastianBoe aligns with my suggestion. As he mentions, for `Kconfig` its unavoidable because we are not invoking the python script from Kconfig. But for CMake it's a completely different story since we are actively calling it from the CMake code.

tejlmand commented:
> > But if it preferred that the CMake output generates a plain file and Kconfig output generates code with osource statements, then I can change.
> 
> Yes, this is preferable.

> I see that @SebastianBoe aligns with my suggestion. As he mentions, for Kconfig its unavoidable because we are not invoking the python script from Kconfig. But for CMake it's a completely different story since we are actively calling it from the CMake code.

Fixed.
Thanks for the feedback.

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+        if section is not None:
+            cmake_path = section.get('cmake', cmake_path)
+            kconfig_file = section.get('kconfig', kconfig_file)
+
```
we should assert when the yaml file is pointing to a non-existent file.

mbolivar replied:
I think you're referring to using try statements to open the file, and doing something like calling `sys.exit` or printing to stderr in the exception handler, right? Not an actual `assert` statement.

tejlmand replied:
@SebastianBoe  yep, see your point, although I agree with @mbolivar  you don't want assert, but instead raise an exception in order to abort with error.

SebastianBoe replied:
Usually I like stack traces, but I see this would be a user-input error through misconfigured .yml, so this should be a message through stderr and an exit code as @mbolivar says.

Internal system error -> stack trace.
User-input error -> error message.

tejlmand replied:
fixed

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+                        help='List of extra modules to parse')
+    args = parser.parse_args()
+
+    if args.kconfig_out is not None:
```
Just

if args.kconfig_out:

is best-practice.

tejlmand replied:
I don't think that is best practice, also PEP8 states:
https://www.python.org/dev/peps/pep-0008/#id51
> Comparisons to singletons like None should always be done with is or is not, never the equality operators.
> 
> Also, beware of writing if x when you really mean if x is not None

And here, I do indeed mean `is not None`

SebastianBoe replied:
> And here, I do indeed mean is not None.

Maybe you do, but we want to express "x is not None and x != empty_string, which we will get with "if x", but we don't get with a None comparison.

Allowing x == empty_string will cause weird issues further down the line AFAICT.

SebastianBoe replied:
@mbolivar : I agree, we wouldn't want to use "if x" for the int type.

tejlmand replied:
@SebastianBoe you're right, haven't consider that a user could write `--cmake-out=` without providing a filename in which case the value would indeed be an empty string.

So that is fixed now.

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+        type: str
+'''
+
+AUTO_GENERATED_HEADER = '# NOTE: THIS FILE IS AUTOGENERATED - DO NOT EDIT\n\n'
```
Can we drop this?

Then we can drop the regex:

https://github.com/zephyrproject-rtos/zephyr/pull/14667/files#diff-af3b638bc2a3e6c650974192a53c7291R581

Users really should know that they can't modify any files in the build directory ...

tejlmand replied:
Fine with me, but it was originally introduced due to this comment: https://github.com/zephyrproject-rtos/zephyr/pull/13067#discussion_r254043635
> - the filename should reflect that, e.g. modules.generated.kconfig
> - Add a header of "# NOTE: THIS IS AUTOGENERATED BY " to this file.

SebastianBoe replied:
I disagree with that comment.

Most files in the build directory are generated (as opposed to copied from source, which is rarely done in CMake-based build systems). And most files do not have the word generated in the filename or in the header.

But I can understand having the disclaimer in generated C code for users with editors that don't make it obvious that they have opened a file in the build directory.

tejlmand replied:
> I disagree with that comment.

Personally this is not important to me, and per default I would not have put it in first place.
I was just referring to why it was there. 

So it's removed again.

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+def generate_file_header(includefile):
+    path = os.path.dirname(includefile)
+    if path != '' and not os.path.exists(path):
+        os.makedirs(path)
```
This can be dropped, the user should be able to make the directory of the path he inputs.

tejlmand replied:
Well, the `user` in this case is CMake, and at this point in time, then the folder `${PROJECT_BINARY_DIR}` doesn't exist.
The reason this worked in CMake was because `file(write <path>/filename)` will automatically create `<path>`, if it doesn't exists.

`${PROJECT_BINARY_DIR}` is used because that variable is already exported to Kconfig

SebastianBoe replied:
CMake should create the directory then.

tejlmand replied:
Folder creation moved to CMake.

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+    path = os.path.dirname(includefile)
+    if path != '' and not os.path.exists(path):
+        os.makedirs(path)
+    if includefile is not None:
```
This if-statement is always true so it can be dropped.

tejlmand replied:
yep. has become obsolete when I refactored code.
Thanks for spotting.

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+    if path != '' and not os.path.exists(path):
+        os.makedirs(path)
+    if includefile is not None:
+        with open(includefile, 'w+') as f:
```
the '+' can be dropped.

tejlmand replied:
agree.
Fixed

SebastianBoe commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+        os.makedirs(path)
+    if includefile is not None:
+        with open(includefile, 'w+') as f:
+            f.write(AUTO_GENERATED_HEADER)
```
Assuming you agree about https://github.com/zephyrproject-rtos/zephyr/pull/14667/files#r266966468 we can just 'pass' here to create the file.

tejlmand replied:
yeah. as I wrote https://github.com/zephyrproject-rtos/zephyr/pull/14667#discussion_r267101171
it seems people have different opinions about headers.

Personally I agree with @SebastianBoe that when files are in build folder, they should be assumed to be generated.
But if people wants a header, that also fine with me.

Purpose of code is correctly to generate the header (and folder if missing)

tejlmand replied:
as it seems we agree to not have the header in files then that code is now removed.

mbolivar commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+    if os.path.isfile(kconfig_file_abs) and kconfig_out is not None:
+        val_str = 'osource "{}"\n\n'.format(kconfig_file_abs)
+
+        with open(kconfig_out, 'a') as f:
```
Same comment for the way `cmake_out` is handled, above, applies here.

mbolivar reviewed (commented):
Just some comments about style since I guess this won't be merged for v1.14 and we are thus in no hurry and can discuss what's Pythonic or not at our leisure ;).

tejlmand commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+    Process a list of projects and create Kconfig / CMake include files for \
+    projects which are also a Zephyr module''')
+
+    parser.add_argument('--kconfig-out', required=False,
```
true, and completely agree.
This was simply reuse of the old code from yaml_to_cmake.py, where the output arg was `required=True`
Should have deleted instead of changing to `False`

tejlmand commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+        val_str = '{}:{}\n'.format(os.path.basename(module), cmake_path_abs)
+
+        with open(cmake_out, 'a') as f:
+            f.write(val_str)
```
yep, good point.
Guess it's just a personal preference for the 
```
with open(<file>) as f:
    f.write(<something>)
```

As you avoid the close, as well as the try-catch for the IO handling.
Although, I do agree, that only opening the file once for writing is a small optimization.

tejlmand commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+
+def main():
+    parser = argparse.ArgumentParser(description='''
+    Process a list of projects and create Kconfig / CMake include files for \
```
no it's not needed.
Wonder how it got in there in the first place.

tejlmand commented on scripts/zephyr_module.py:
```diff
@@ -0,0 +1,145 @@
+            meta = yaml.safe_load(f.read())
+
+        schema = yaml.safe_load(METADATA_SCHEMA)
+        pykwalify.core.Core(source_data=meta, schema_data=schema).validate()
```
guess moving code around catches you attention ;)

tejlmand replied:
now catching a printing more informative message to user.

tejlmand commented:
@mbolivar @SebastianBoe all comments addressed

Also removed the `section` from arguments to be internal knowledge, as that is where it correctly belongs.
The `section` was leftover from the original `yaml_to_cmake.py` script

tejlmand commented:
Thanks for the review

tejlmand commented:
@carlescufi @SebastianBoe @mbolivar Can you recheck, everything should be fixed :)

carlescufi reviewed (approved):
Looks good to me now, thank you!

This will need small fixes after https://github.com/zephyrproject-rtos/zephyr/pull/14976 is merged.

carlescufi committed:
doc: Process zephyr modules to parse all of Kconfig

In order to parse all the Kconfig files that make up the whole Kconfig
tree we need to process the modules, since this can optionally provide
Kconfig files. In order to do so, include the modules CMake file so that
they are listed and a Kconfig.modules is generated.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
doc: Enable undef warnings in Kconfig

When generating the Kconfig reference documentation, enable the warnings
for undefined Kconfig options in order to make sure all references are
met.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented:
@nashif @mbolivar @tejlmand pushed a couple of commits to fix #14989

carlescufi commented:
@mbolivar @nashif @tejlmand this is ready to be merged

carlescufi commented:
@tejlmand @nashif after this is merged, we should remove the equivalent python code in `ci-tools/scripts/check_compliance.py` and use this `zephyr_module.py` script instead.

marc-hb commented:
Looks like this broke "hello_world" (and everything else) when not using west. I'm aware not using west is "living in the past", however I didn't really expect this at this time between -rc3 and -rc4. A meaningful error message mentioning (the lack of) west would also be nicer than a cryptic kconfig stack trace. Has the documentation been updated to clarify that west is now mandatory?

> Just some comments about style since I guess this won't be merged for v1.14 and we are thus in no hurry 

Famous last words :-)

ioannisg commented:
@carlescufi @tejlmand @nashif 
Merging this PR has broken my Windows10 build setup:
Here's some log info, when running` cmake -GNinja -DBOARD=nrf52840_pca10056 ..`:


```

-- The C compiler identification is GNU 7.3.1
-- The CXX compiler identification is GNU 7.3.1
-- The ASM compiler identification is GNU
-- Found assembler: C:/gccarmemb/bin/arm-none-eabi-gcc.exe
-- Performing Test toolchain_is_ok
-- Performing Test toolchain_is_ok - Success
Including module: tinycbor:C
CMake Warning at ../../../../CMakeLists.txt:1464 (message):


        ------------------------------------------------------------
        --- WARNING:  __ASSERT() statements are globally ENABLED ---
        --- The kernel will run more slowly and use more memory  ---
        ------------------------------------------------------------


-- Configuring done
CMake Error: Cannot open file for write: C:/Users/iogl/Documents/nordic/workspace/zephyr/zephyr/tests/kernel/mem_protect/userspace/build_nrf52840_pca10056_2/tinycbor:C/cmake_install.cmake.tmp
CMake Error: : System Error: Invalid argument
```

Looks like we are trying to include module `tinycbor:C`

I am no expert on this, but reverting these commits seem to solve the problem.

aescolar commented:
@ioannisg : Does #15042 fix it for you? Could you or @marc-hb file a bug for this regression?

carlescufi commented:
This is very strange. A module named "tinycbor:C" is being found by west. Can you please run the following `west list -format={posixpath}`?
